### PR TITLE
Make shell scripts POSIX-compatible

### LINKED ISSUE #3544: Make shell scripts POSIX-compatible
Some environments don't have Bash, and all PHP-CS-Fixer's shell scripts fail with

> env: can't execute 'bash': No such file or directory

It would be great if PHP-CS-Fixer's shell scripts were POSIX-compatible.

### LINKED ISSUE #3613: Add ShellCheck to CI
[ShellCheck](https://www.shellcheck.net/) is a great static analyzer for shell scripts.

I propose to add it to CI.

### PR DESCRIPTION
**Pull request by EvgenyOrekhov**
I used ShellCheck and checkbashisms to find and get rid of Bash-specific syntax and features in shell scripts.

This PR is not about refactoring or fixing the scripts. The intention of this PR is to make the scripts work in any POSIX-compliant shell.

This will allow developers to contribute to PHP-CS-Fixer using environments that don't have Bash, e. g. Alpine Linux.

Fixes #3544.
Fixes #3613.
Closes #3609.

### CODE REVIEW

File: benchmark.sh [hunk 1]
```diff
@@ -1,8 +1,8 @@
 #!/bin/sh
 
-command -v php >/dev/null 2>&1 || { echo "I require `php` but it's not available. Aborting." >&2; exit 255; }
```
Comment (keradus) [inline benchmark.sh]: `command -v php` will get path of php,
`command php` will execute php

```
ker@dus:~ λ command -v php
/usr/bin/php
ker@dus:~ λ command php
^C # new hanging out process of PHP was created and it awaits for input, closing with CTRL+C
```

File: .travis.yml [hunk 2]
```diff
@@ -62,6 +58,7 @@ jobs:
                 - ./check_trailing_spaces.sh || travis_terminate 1
                 - if [ -n "$COMMIT_SCA_FILES" ]; then ./dev-tools/vendor/bin/phpmd `echo "$COMMIT_SCA_FILES" | grep -Ev "^tests/Fixtures" | xargs | sed 's/ /,/g'` text phpmd.xml || travis_terminate 1; fi
                 - ./dev-tools/vendor/bin/composer-require-checker check composer.json --config-file=.composer-require-checker.json
+                - ./dev-tools/check-shell-scripts.sh
```
Comment (keradus) [inline .travis.yml]:
```
script ./dev-tools/vendor/nikic/php-parser/test_old/run-php-src.sh does not appear to have a #! interpreter line;
you may get strange results
```

wow, we shall not check the vendor

File: benchmark.sh [hunk 3]
```diff
@@ -1,21 +1,21 @@
-#!/bin/bash
+#!/bin/sh
 
-command -v php >/dev/null 2>&1 || { echo "I require `php` but it's not available. Aborting." >&2; exit 255; }
-command -v grep >/dev/null 2>&1 || { echo "I require `grep` but it's not available. Aborting." >&2; exit 255; }
-command -v awk >/dev/null 2>&1 || { echo "I require `awk` but it's not available. Aborting." >&2; exit 255; }
+command php >/dev/null 2>&1 || { echo "I require `php` but it's not available. Aborting." >&2; exit 255; }
```
Comment (keradus) [inline benchmark.sh]: `command -v php` will get path of php,
`command php` will execute php

```
ker@dus:~ λ command -v php
/usr/bin/php
ker@dus:~ λ command php
^C # new hanging out process of PHP was created and it awaits for input, closing with CTRL+C
```

File: .travis.yml [hunk 4]
```diff
@@ -2,6 +2,11 @@ language: php
 
 sudo: false
 
+addons:
+    apt:
+        packages:
+            - devscripts
```
Comment (keradus) [inline .travis.yml]: :-1:

Comment (EvgenyOrekhov) [reply .travis.yml]: @keradus I found a more recent version of `checkbashisms`. Its home page is located here: https://sourceforge.net/projects/checkbaskisms/. `checkbashisms` is just a Perl script that can be downloaded using `wget` and put into `/usr/bin`. How would you recommend me to install it for CI?

Comment (keradus) [reply .travis.yml]: either install `checkbashisms` as apt, similar to like you do it now, or manually. both are fine for me, as long as you do it only for the only job that needs it.
Currently, you install it again and again for every single job we have.

**Comment by keradus** [comment]
@EvgenyOrekhov , any update on this one ? :)

**Comment by EvgenyOrekhov** [comment]
@keradus I will try to rewrite CiIntegrationTest and ci-integration.sh when I have time.

**Comment by keradus** [comment]
does last comment means you were already able to fix ci-integration and all posix-comaptible is changes are ready?
sorry for delay, but last notification I got for this PR was your comment, not commit.

if so, please, could you fix up Travis build?

**Comment by EvgenyOrekhov** [comment]
@keradus Yes, I made all the scripts POSIX-compatible, but I am unable to fix the build due to a bug in README generation command. When I run `php php-cs-fixer readme > README.rst` it generates a trailing space on one of the lines, as you can see here: https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/352442019#L647. If I remove that trailing space, then the build fails saying that I need to regenerate README.

**Comment by keradus** [comment]
why readme generation doesn't produce trailing whitespaces on 2.2 and does on your branch ?

**Comment by EvgenyOrekhov** [comment]
@keradus Because I changed ci-integration.sh, and it gets copy-pasted to README.rst during generation. But my version of ci-integration.sh doesn't contain trailing spaces. This is weird.

File: .travis.yml [hunk 5]
```diff
@@ -33,6 +33,9 @@ jobs:
 
                 - travis_retry composer update $DEFAULT_COMPOSER_FLAGS $COMPOSER_FLAGS
                 - composer info -D | sort
+
+                - wget --output-document=$HOME/bin/checkbashisms https://sourceforge.net/projects/checkbaskisms/files/2.0.0.2/checkbashisms/download
```
Comment (keradus) [inline .travis.yml]: put it into cache directory and add `-nc`, that way we won't install it every single job

File: dev-tools/check-shell-scripts.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+set -eu
```
Comment (keradus) [inline dev-tools/check-shell-scripts.sh]: :+1: for `u`

File: dev-tools/check-shell-scripts.sh [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+IFS='
+'
+
+SHELL_SCRIPTS=$(find ./ -type f -not -path '*/vendor/*' -iname '*.sh')
+
+# shellcheck disable=SC2086
```
Comment (keradus) [inline dev-tools/check-shell-scripts.sh]: is this comment do anything (some bashism magic) or could be removed?

File: dev-tools/ci-integration.sh [hunk 8]
```diff
@@ -1,6 +1,9 @@
-#!/usr/bin/env bash
+#!/bin/sh
 set -e
 
+IFS='
```
Comment (keradus) [inline dev-tools/ci-integration.sh]: btw, that is the benefit of moving it out of the command, near `EXTRA_ARGS` ?

**Comment by EvgenyOrekhov** [comment]
@keradus I added `$HOME/bin` to cache, and I added `--no-clobber` to `wget` in .travis.yml.

Sorry if I messed up the conversation. I can't figure out how to properly use GitHub reviews/comments.

**Comment by EvgenyOrekhov** [comment]
@keradus It was proposed [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3546#issuecomment-366521838), [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3609#issuecomment-374119509), and [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3613#issuecomment-374263680) to add ShellCheck to CI in this PR.

I can do that, but it will make this PR's scope broader (making the shell scripts POSIX-compatible + refactoring), which, IMHO, not very good.

What do you say? Should I add ShellCheck to CI and lint the shell scripts according to its warnings in this PR?

**Comment by EvgenyOrekhov** [comment]
@keradus I went ahead and added ShellCheck to CI and linted all shell scripts.

This PR now fixes #3544, #3613, and closes #3609.

**Comment by EvgenyOrekhov** [comment]
@keradus `wget` always exits with code `1` when used with `--no-clobber` and `--output-document` and the file exists: https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/355473873#L566

Should I just add `|| true` after `wget` to ignore the error?

File: dev-tools/check-shell-scripts.sh (hunk 7 above)
Comment (keradus) [reply dev-tools/check-shell-scripts.sh]: ok, now I got lost.
fixing ShellCheck without adding it to CI will make the violations to appear again soon

Comment (EvgenyOrekhov) [reply dev-tools/check-shell-scripts.sh]: I added ShellCheck to CI. You can see it on line 14 in this file.

File: README.rst [hunk 9]
```diff
@@ -1237,9 +1237,12 @@ Then, add the following command to your CI:
 
 .. code-block:: bash
 
+    $ IFS='
+    $ '
     $ CHANGED_FILES=$(git diff --name-only --diff-filter=ACMRTUXB "${COMMIT_RANGE}")
-    $ if ! echo "${CHANGED_FILES}" | grep -qE "^(\\.php_cs(\\.dist)?|composer\\.lock)$"; then IFS=$'\n' EXTRA_ARGS=('--path-mode=intersection' '--' ${CHANGED_FILES[@]}); fi
-    $ vendor/bin/php-cs-fixer fix --config=.php_cs.dist -v --dry-run --stop-on-violation --using-cache=no "${EXTRA_ARGS[@]}"
+    $ if ! echo "${CHANGED_FILES}" | grep -qE "^(\\.php_cs(\\.dist)?|composer\\.lock)$"; then EXTRA_ARGS=$(printf -- '--path-mode=intersection\n--\n%s' "${CHANGED_FILES}"); fi
+    $ # shellcheck disable=SC2086
```
Comment (keradus) [inline README.rst]: this is snippet of what end-user could put to his CI script.
this comment is not interesting for him for sure

Comment (EvgenyOrekhov) [reply README.rst]: I replaced that comment with a command line flag in check-shell-scripts.sh: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3548/commits/b4809150fc1dfcf8093f39c028944843ae856af8

**Comment by EvgenyOrekhov** [comment]
@keradus I think I'm finished. Please review.

File: dev-tools/ci-integration.sh [hunk 10]
```diff
@@ -1,6 +1,8 @@
...
```
Comment (keradus) [inline dev-tools/ci-integration.sh]: just wondering... would
`printf -v IFS "\n"`
work ?

File: dev-tools/check-shell-scripts.sh [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+SHELL_SCRIPTS=$(
+    find ./ \
+    -type f \
+    -not -path '*/vendor/*' \
+    -not -path './dev-tools/ci-integration.sh' \
+    -iname '*.sh'
+)
+
+# shellcheck disable=SC2086
+checkbashisms $SHELL_SCRIPTS
```
Comment (keradus) [inline dev-tools/check-shell-scripts.sh]: is `"${SHELL_SCRIPTS[@]}"` valid in `sh` (i know it is in `bash`) ?

File: dev-tools/ci-integration.sh (hunk 10 above)
Comment (EvgenyOrekhov) [reply dev-tools/ci-integration.sh]: `printf -v` is a bashism.

You mean `$'...'` as in `IFS=$'\n'`? No, `$'...'` is also a bashism.

File: dev-tools/check-shell-scripts.sh (hunk 11 above)
Comment (EvgenyOrekhov) [reply dev-tools/check-shell-scripts.sh]: No. There are no user-defined arrays in POSIX, there is only one quasi-array `$@` (the arguments array). That's why I had to replace all arrays with strings and use `IFS` to split them.

File: check_trailing_spaces.sh [hunk 12]
```diff
@@ -1,14 +1,14 @@
...
 files_with_trailing_spaces=$(find . -type f -not -path "./.git/*" -not -path "./dev-tools/vendor/*" -not -path "./vendor/*" -not -path "./tests/Fixtures/*" -exec egrep -nH " $" {} \;)
 
-if [[ $files_with_trailing_spaces ]]
+if [ "$files_with_trailing_spaces" ]
 then
-    echo -e "\e[97;41mTrailing spaces detected:\e[0m"
+    printf '\033[97;41mTrailing spaces detected:\033[0m\n'
     e=$(printf '\033')
-    echo "${files_with_trailing_spaces}" | sed -E "s/^\.\/([^:]+):([0-9]+):(.*[^ ])( +)$/${e}[0;31m - in ${e}[0;33m\1${e}[0;31m at line ${e}[0;33m\2\n   ${e}[0;31m>${e}[0m \3${e}[41m\4${e}[0m/"
+    echo "${files_with_trailing_spaces}" | sed -E "s/^\\.\\/([^:]+):([0-9]+):(.*[^ ])( +)$/${e}[0;31m - in ${e}[0;33m\\1${e}[0;31m at line ${e}[0;33m\\2\\n   ${e}[0;31m>${e}[0m \\3${e}[41m\\4${e}[0m/"
```
Comment (keradus) [inline check_trailing_spaces.sh]: I copy-pasted this script to upstream:
https://github.com/PHP-CS-Fixer/trailing-whitespace-checker/pull/8/files

with changes in this line, tests on upstream repo would be failing. could you give it a try ?

Comment (julienfalque) [reply check_trailing_spaces.sh]: Actually tests pass. There was one weird Travis failure but I restarted it and it passed.

Comment (keradus) [reply check_trailing_spaces.sh]: nope, take a look, I copied whole file except this one line.

Comment (julienfalque) [reply check_trailing_spaces.sh]: Ah sorry, didn't notice that.

File: check_trailing_spaces.sh [hunk 13]
```diff
@@ -15,7 +15,7 @@ if [ "$files_with_trailing_spaces" ]
 then
     printf '\033[97;41mTrailing whitespaces detected:\033[0m\n'
     e=$(printf '\033')
-    echo "${files_with_trailing_spaces}" | sed -E "s/^\\.\\/([^:]+):([0-9]+):(.*[^\t ])?([\t ]+)$/${e}[0;31m - in ${e}[0;33m\\1${e}[0;31m at line ${e}[0;33m\\2\\n   ${e}[0;31m>${e}[0m \\3${e}[41;1m\\4${e}[0m/"
+    echo "${files_with_trailing_spaces}" | sed -E "s/^\\.\\/([^:]+):([0-9]+):(.*[^\\t ])?([\\t ]+)$/${e}[0;31m - in ${e}[0;33m\\1${e}[0;31m at line ${e}[0;33m\\2\\n   ${e}[0;31m>${e}[0m \\3${e}[41;1m\\4${e}[0m/"
```
Comment (keradus) [inline check_trailing_spaces.sh]: please, raise this change to https://github.com/PHP-CS-Fixer/trailing-whitespace-checker and let us see what would tests for it say ;)

Comment (EvgenyOrekhov) [reply check_trailing_spaces.sh]: @keradus The tests pass: https://github.com/PHP-CS-Fixer/trailing-whitespace-checker/pull/10

Comment (keradus) [reply check_trailing_spaces.sh]: sounds great, original regex at PHP CS Fixer, that got fixed in your PR initially, was slightly different, and fixed version of it makes tests not pass, with new regex, it's passing indeed. great !

**Comment by keradus** [comment]
Thank you @EvgenyOrekhov.

**Pull request closed without merging**